semgrep-core/src/engine/Selector.ml

```diff
@@ -0,0 +1,130 @@
+module R = Rule
+module AR = Augmented_range
+module G = AST_generic
+module M = Metavariable
+module PM = Pattern_match
+module RP = Report
+
+type selector = {
+  mvar : M.mvar;
+  pid : int;
+  pstr : string;
+  lazy_matches : RP.times RP.match_result Lazy.t;
+}
+
+type sformula =
+  | Leaf of R.leaf
+  | And of (selector option * sformula list)
+  | Or of sformula list
+  (* There are restrictions on where a Not can appear in a formula. It
+   * should always be inside an And to be intersected with "positive" formula.
+   *)
+  | Not of sformula
+
+(*****************************************************************************)
+(* Selecting methods *)
+(*****************************************************************************)
+
+(* this will be adjusted later in range_to_pattern_match_adjusted *)
+let fake_rule_id (id, str) =
+  { PM.id = string_of_int id; pattern_string = str; message = "" }
+
+let match_selector (sel_opt : selector option) : AR.ranges =
+  let get_match (x : RP.times RP.match_result) = x.matches in
+  match sel_opt with
+  | None -> failwith "empty And; no positive terms in And"
```

Comment:
This error message has been taken out of context, so it doesn't make a lot of sense here. Why should we fail with `"empty And; ..."` if `sel_opt` is `None` ? This function could instead return `AR.ranges option` and we could raise the error in `evaluate_formula`.